specifyweb/frontend/js_src/lib/components/DataModel/uniquenessRules.ts

```diff
@@ -0,0 +1,165 @@
+  readonly fields: RA<
+    Pick<SerializedResource<SpLocaleContainerItem>, 'id' | 'name'>
+  >;
+  readonly scopes: RA<
+    Pick<SerializedResource<SpLocaleContainerItem>, 'id' | 'name'>
+  >;
+  readonly isDatabaseConstraint: boolean;
+};
+
+export type UniquenessRules = Partial<
+  RR<
+    keyof Tables,
+    | RA<{
+        readonly rule: UniquenessRule;
+        readonly duplicates?: UniquenessRuleValidation;
+      }>
+    | undefined
+  >
+>;
+
+export type UniquenessRuleValidation = {
+  readonly totalDuplicates: number;
+  readonly fields: RA<{
+    readonly fields: IR<string>;
+    readonly duplicates: number;
+  }>;
+};
+
+export const databaseFieldName = '_database';
+export const databaseScope: SerializedResource<SpLocaleContainerItem> &
+  WithFetchedStrings = {
+  name: databaseFieldName,
+  strings: { name: { text: schemaText.database() } },
+};
+
+let uniquenessRules: UniquenessRules = {};
+
+export const fetchContext = import('./schema')
+  .then(async ({ fetchContext }) => fetchContext)
+  .then(async (schema) =>
+    load<UniquenessRules>(
+      `/businessrules/uniqueness_rules/${schema.domainLevelIds.discipline}/`,
+      'application/json'
+    )
+  )
+  .then((data) =>
+    Object.fromEntries(
+      Object.entries(data).map(([lowercaseTableName, rules]) => [
+        // Convert all lowercase table names from backend to PascalCase
+        strictGetTable(lowercaseTableName).name,
+        rules?.map(({ rule }) => ({
+          rule: {
+            ...rule,
+            scopes: rule.scopes.length === 0 ? [] : rule.scopes,
+          },
+        })),
+      ])
+    )
+  )
+  .then((data) => {
+    uniquenessRules = data;
+    return data;
+  });
+
+export function getUniquenessRules(): UniquenessRules | undefined;
+export function getUniquenessRules<TABLE_NAME extends keyof Tables>(
+  table: TABLE_NAME
+): UniquenessRules[TABLE_NAME] | undefined;
+export function getUniquenessRules<TABLE_NAME extends keyof Tables>(
+  table?: TABLE_NAME
+): UniquenessRules | UniquenessRules[TABLE_NAME] {
+  return Object.keys(uniquenessRules).length === 0
+    ? undefined
+    : table === undefined
+    ? uniquenessRules
+    : uniquenessRules[table];
+}
+
+export function useTableUniquenessRules(
+  tableName: keyof Tables
+): readonly [
+  ...tableRules: GetOrSet<UniquenessRules[keyof Tables]>,
+  setCachedTableRules: (value: UniquenessRules[keyof Tables]) => void
+] {
+  const [rawModelRules = [], setTableUniquenessRules] = React.useState(
+    uniquenessRules[tableName]
+  );
+
+  const setStoredUniquenessRules = (
+    value: UniquenessRules[keyof Tables]
+  ): void => {
+    uniquenessRules = Object.fromEntries(
+      Object.entries(uniquenessRules).map(([table, rules]) => [
+        table,
+        table === tableName ? value : rules,
+      ])
+    );
+  };
+
+  const modelRules = React.useMemo(() => rawModelRules, [rawModelRules]);
+
+  return [modelRules, setTableUniquenessRules, setStoredUniquenessRules];
```

Comment:
This can be simplified. 
```suggestion
  return [rawModelRules, setTableUniquenessRules, setStoredUniquenessRules];
```

Reply:
Fixed in 995d822cadf12acd3b6a1d3893570b3044a17ddc